internal/querycoordv2/meta/target_manager.go

```diff
@@ -3,184 +3,297 @@ package meta
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
+	log.Info("start to update current target for partition",
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("PartitionIDs", partitionIDs))
+
+	newTarget := mgr.next.getCollectionTarget(collectionID)
+
+	if newTarget == nil || newTarget.isEmpty() {
+		log.Info("next target does not exist, skip it",
+			zap.Int64("collectionID", collectionID),
+			zap.Int64s("PartitionIDs", partitionIDs))
+		return
 	}
+	mgr.current.updateCollectionTarget(collectionID, newTarget)
+	mgr.next.removeCollectionTarget(collectionID)
+
+	log.Info("finish to update current target for partition",
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("PartitionIDs", partitionIDs),
+		zap.Int64s("segments", newTarget.GetAllSegmentIDs()),
+		zap.Strings("channels", newTarget.GetAllDmChannelNames()))
+}
+
+func (mgr *TargetManager) UpdateCollectionNextTarget(collectionID int64, newTarget *CollectionTarget) error {
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
+	log.Info("start to update next targets for collection",
+		zap.Int64("collectionID", collectionID))
+
+	mgr.next.updateCollectionTarget(collectionID, newTarget)
+
+	log.Info("finish to update next targets for collection",
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("segments", newTarget.GetAllSegmentIDs()),
+		zap.Strings("channels", newTarget.GetAllDmChannelNames()))
+	return nil
 }
 
 // RemoveCollection removes all channels and segments in the given collection
 func (mgr *TargetManager) RemoveCollection(collectionID int64) {
-	mgr.rwmutex.Lock()
-	defer mgr.rwmutex.Unlock()
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
+	log.Info("remove collection from targets",
+		zap.Int64("collectionID", collectionID))
 
-	log.Info("remove collection from targets")
-	for _, segment := range mgr.segments {
-		if segment.CollectionID == collectionID {
-			mgr.removeSegment(segment.GetID())
-		}
-	}
-	for _, dmChannel := range mgr.dmChannels {
-		if dmChannel.CollectionID == collectionID {
-			mgr.removeDmChannel(dmChannel.GetChannelName())
-		}
-	}
+	mgr.current.removeCollectionTarget(collectionID)
+	mgr.next.removeCollectionTarget(collectionID)
 }
 
 // RemovePartition removes all segment in the given partition,
 // NOTE: this doesn't remove any channel even the given one is the only partition
-func (mgr *TargetManager) RemovePartition(partitionID int64) {
-	mgr.rwmutex.Lock()
-	defer mgr.rwmutex.Unlock()
-
+func (mgr *TargetManager) RemovePartition(collectionID int64, partitionIDs ...int64) {
+	mgr.rwMutex.Lock()
+	defer mgr.rwMutex.Unlock()
 	log.Info("remove partition from targets",
-		zap.Int64("partitionID", partitionID))
-	for _, segment := range mgr.segments {
-		if segment.GetPartitionID() == partitionID {
-			mgr.removeSegment(segment.GetID())
+		zap.Int64("collectionID", collectionID),
+		zap.Int64s("partitionIDs", partitionIDs))
+
+	partitionSet := typeutil.NewUniqueSet(partitionIDs...)
+
+	currentTarget := mgr.getTarget(CurrentTarget)
```

Comment:
Why not use mgr.current directly?

Reply:
it seems ugly to judge current type everywhere, so make it a function call for logic reuse

Reply:
> it seems ugly to check target type everywhere, so make it a function call for logic reuse

But this function is just to fetch current target here. And you can get current by `mgr.current` like line83.

Reply:
agree